presidio-image-redactor/tests/test_dicom_image_pii_verify_engine.py

```diff
@@ -0,0 +1,929 @@
+    )
+
+    # Assert
+    assert test_all_pos == expected_results
+
+
+# ------------------------------------------------------
+# DicomImagePiiVerifyEngine._calculate_precision()
+# ------------------------------------------------------
+@pytest.mark.parametrize(
+    "gt_labels, all_pos, expected_result",
+    [
+        (
+            {"0": {}, "1": {}, "2": {}, "3": {}},
+            {"0": {}, "1": {}, "2": {}, "3": {}},
+            1.0,
+        ),
+        ({"0": {}, "1": {}, "2": {}, "3": {}}, {"0": {}, "1": {}, "2": {}}, 1.0),
+        ({"0": {}, "1": {}, "2": {}, "3": {}}, {"0": {}, "1": {}}, 1.0),
+        ({"0": {}, "1": {}, "2": {}, "3": {}}, {"0": {}}, 1.0),
+        ({"99": {}}, {"0": {}, "1": {}, "2": {}, "3": {}}, 0.0),
+        ({"0": {}}, {"0": {}, "1": {}, "2": {}, "3": {}}, 0.25),
+        ({"0": {}, "1": {}}, {"0": {}, "1": {}, "2": {}, "3": {}}, 0.5),
+        ({"0": {}, "1": {}, "2": {}}, {"0": {}, "1": {}, "2": {}, "3": {}}, 0.75),
+    ],
+)
+def test_calculate_precision_happy_path(
+    mock_engine: DicomImagePiiVerifyEngine,
+    gt_labels: dict,
+    all_pos: dict,
+    expected_result: float,
+):
+    """Test happy path for DicomImagePiiVerifyEngine._calculate_precision
+
+    Args:
+        mock_engine (DicomImagePiiVerifyEngine): Instantiated engine.
+        gt_labels (dict): Ground truth labels for single instance.
+        all_pos (dict): All positives (matched detected PHI).
+        expected_result (float): Expected precision.
+    """
+    # Act
+    test_precision = mock_engine._calculate_precision(gt_labels, all_pos)
+
+    # Assert
+    assert test_precision == expected_result
+
+
+# ------------------------------------------------------
+# DicomImagePiiVerifyEngine._calculate_recall()
+# ------------------------------------------------------
+@pytest.mark.parametrize(
+    "gt_labels, all_pos, expected_result",
+    [
+        (
+            {"0": {}, "1": {}, "2": {}, "3": {}},
+            {"0": {}, "1": {}, "2": {}, "3": {}},
+            1.0,
+        ),
+        ({"0": {}, "1": {}, "2": {}, "3": {}}, {"0": {}, "1": {}, "2": {}}, 0.75),
+        ({"0": {}, "1": {}, "2": {}, "3": {}}, {"0": {}, "1": {}}, 0.5),
+        ({"0": {}, "1": {}, "2": {}, "3": {}}, {"0": {}}, 0.25),
+        ({"0": {}, "1": {}, "2": {}, "3": {}}, {"0": {}, "999": {}}, 0.25),
+    ],
+)
+def test_calculate_recall_happy_path(
+    mock_engine: DicomImagePiiVerifyEngine,
+    gt_labels: dict,
+    all_pos: dict,
+    expected_result: float,
+):
+    """Test happy path for DicomImagePiiVerifyEngine._calculate_recall
+
+    Args:
+        mock_engine (DicomImagePiiVerifyEngine): Instantiated engine.
+        gt_labels (dict): Ground truth labels for single instance.
+        all_pos (dict): All positives (matched detected PHI).
+        expected_result (float): Expected precision.
+    """
+    # Act
+    test_recall = mock_engine._calculate_recall(gt_labels, all_pos)
```

Comment:
Yes, I test private methods as well considering they are code I write which could have mistakes by accident if not tested.

The test for `_calculate_recall` is the last test in this file.

Reply:
Personally, I am not testing private methods. The reasoning behind it is that only public methods are accessible to a client and act as the contract with it. Altering them is considered as a breaking change, but the inner implementation and the use of private methods can change without affecting the functionality. IMO it should be tested the same, meaning changing privates should not affect or break tests (clients) but changing public should. Natively, is some languages, privates are not even accessible and can't be mocked without special libraries (Java needs reflection for example)

Reply:
I see your point, thank you for explaining! However, I still feel that having these tests are really useful.

I have been using a test-driven development approach for writing these private methods. Having the unit tests for private methods was critical in saving me time when I was updating the logic of existing code to use lists instead of dictionaries in this PR.

To your point about developers needing to alter private methods without causing breaking changes, I feel the developers should have a good enough understanding of any methods they are altering to be able to update the associated unit tests as necessary.

Also, although these methods do have the underscore prefix, they are not truly private methods.

Reply:
I don't have a strong preference here :) @niwilso if we could think of test cases to test private methods through public methods, and make sure they work as expected, that would be the best, but it's not always straightforward. Another option is to consider making those public. Maybe a user would want access to precision/recall methods?

Reply:
Ah good point. I'll make the precision and recall calculation methods public since there are cases where the user may want to call them.

I'll revisit testing private methods through public methods in the future PR for refactoring the bounding box code into a common module.